### PR TITLE
[POC] Add AMP zero gradients optimization

### PR DESCRIPTION
**Pull request by ymwangg**
This PR is the POC of the AMP zero gradients optimization.

### CODE REVIEW

**Comment by JackCaoG** [comment]
Sorry for the delay, will try to take a look tmr.

**Review by JackCaoG — CHANGES_REQUESTED** [review]
Thanks for the PR, some questions and minor comments.

File: test/test_train_mp_imagenet.py [hunk 1]
```diff
@@ -223,7 +228,6 @@ def train_loop_fn(loader, epoch):
         xm.all_reduce('sum', gradients, scale=1.0 / xm.xrt_world_size())
         scaler.step(optimizer)
         scaler.update()
-        xm.mark_step()
```
[the remark this answers is not in the document]
Comment (ymwangg) [reply test/test_train_mp_imagenet.py:226]: I found this `mark_step` is redundant since the script is using `pl.MpDeviceLoader` and it diminishes the performance gain of the `zero_gard` optimization.

File: torch_xla/amp/grad_scaler.py [hunk 2]
```diff
@@ -1,13 +1,74 @@
...
+      enabled (bool, optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+          invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+      use_zero_grad (bool, optional, default=False): If ``True``, enables the torch_xla specific zero gradients
+          optimization that performs ``optimizer.step()`` with gradients set to zero instead of skipping it when
+          inf/NaN gradients occur. This may improve the performance by removing the barrier in GradScaler.
+  """
+
+  def __init__(
+      self,
+      init_scale=2.0**16,
```
[the remark this answers is not in the document]
Comment (ymwangg) [reply torch_xla/amp/grad_scaler.py:29]: It looks like `2.0**16 == 65536.0`. This essentially copied`torch.cuda.amp.GradScaler.__init__` here https://github.com/pytorch/pytorch/blob/master/torch/cuda/amp/grad_scaler.py#L108-L113 with an extra flag.

File: test/test_train_mp_imagenet.py [hunk 3]
```diff
@@ -206,7 +210,8 @@ def train_imagenet():
       summary_writer=writer)
   loss_fn = nn.CrossEntropyLoss()
   if FLAGS.amp:
-    scaler = GradScaler()
+    scaler = GradScaler() if not FLAGS.use_zero_grad else GradScaler(
+        use_zero_grad=True)
```
[the remark this answers is not in the document]
Comment (ymwangg) [reply test/test_train_mp_imagenet.py]: good suggestion!

File: torch_xla/amp/grad_scaler.py [hunk 4]
```diff
@@ -1,13 +1,74 @@
...
+
+    def get_scaling_factor(a):
+
+      def if_true(a):
+        return xb.Op.zero(a.builder())
+
+      def if_false(a):
+        return xb.Op.one(a.builder())
+
+      cond = a > xb.Op.zero(a.builder())
```
Comment (JackCaoG) [inline torch_xla/amp/grad_scaler.py:52]: Will `a != xb.Op.zero(a.builder())` be a better equivalent of `if not sum(....)`

**Review by JackCaoG — CHANGES_REQUESTED** [review]
Mostly LGTM, one minor nit

File: torch_xla/amp/grad_scaler.py (hunk 4 above)
Comment (ymwangg) [reply torch_xla/amp/grad_scaler.py]: Yes, I agree.

**Review by JackCaoG — APPROVED** [review]
Thanks! I will merge the change once all test passed. Could you share the speed data(which model, what batch size, how long it takes) on the RFC you filed so it has a bit more info?

**Comment by ymwangg** [comment]
> Thanks! I will merge the change once all test passed. Could you share the speed data(which model, what batch size, how long it takes) on the RFC you filed so it has a bit more info?

Thanks, will update benchmark numbers in the RFC shortly.

**Comment by ymwangg** [comment]
> > Thanks! I will merge the change once all test passed. Could you share the speed data(which model, what batch size, how long it takes) on the RFC you filed so it has a bit more info?
> 
> Thanks, will update benchmark numbers in the RFC shortly.

Updated benchmark results of MNIST and ImageNet test cases in the original RFC https://github.com/pytorch/xla/issues/3120

**Comment by JackCaoG** [comment]
@ymwangg Benchmark looks great! Thanks for sharing that.
